File: qqplot.jl
#!/usr/bin/env julia

using ArgParse
using StatsPlots
using CodecZlib
using DelimitedFiles
using Dates

# disable calling display window
ENV["GKSwstype"] = "100"

function qqfly(out::String, title::String="", cutoff::Float64=1e-4, bins::Int=1000; cmd=nothing)
    println(Dates.format(now(), "YY-mm-dd:HH:MM:SS  "), "program started.")
    allbins = [Float64[-1, 0] for i in 1:bins]
    sigp = Float64[]
    N = 0
    c = -log10(cutoff)
    if cmd == nothing
        io = stdin
    else
        io = open(cmd)
    end
    while !eof(io)
        N += 1
        if N % 100000000 == 0
            println(Dates.format(now(), "YY-mm-dd:HH:MM:SS  "), "reached the $N line.")
        end

        p = -log10(parse(Float64, readline(io)))
        if p > c
            sigp = [sigp; p]
        else
            i = bins - Int(floor(p / c * bins))
            i == 0 ? i = 1 : i
            allbins[i][1] = max(allbins[i][1], p)
            allbins[i][2] += 1
        end
    end
    println(Dates.format(now(), "YY-mm-dd:HH:MM:SS  "), "total number of lines is $N.")

    size = length(sigp)
    obs = sort(sigp, rev=true)
    exp = [-log10((i + 1 - 0.5) / N) for i in 1:size]
    for b in allbins
        if b[1] != -1
            obs = [obs; b[1]]
            exp = [exp; -log10((b[2] + size - 0.5) / N)]
            size += b[2]
        end
    end
    # plotting
    gr(size=(1000, 1000))
    qqplot(exp, obs, xlabel="Expected -log10(P)", ylabel="Observed -log10(P)", markerstrokecolor=:deepskyblue3, linecolor=:red, markercolor=:deepskyblue3, linewidth=3, title=title)
    figname = out * ".png"
    savefig(figname)
    csvname = out * ".csv.gz"
    open(GzipCompressorStream, csvname, "w") do stream
        writedlm(stream, [obs exp], ',')
    end
    println(Dates.format(now(), "YY-mm-dd:HH:MM:SS  "), "program finished.")

    return nothing
end


function main()
    s = ArgParseSettings()
    @add_arg_table s begin
        "--out"
            help = "prefix of output"
            default = "qqplot-jl"
        "--title"
            help = "title of the plot"
            default = ""
        "--cutoff"
            help = "pval bigger than this cutoff will be grouped into bins.[1e-4]"
            arg_type = Float64
            default = 1e-4
        "--bins"
            help = "the number of bins to use.[1000]"
            arg_type = Int
            default = 1000
    end

    args = parse_args(s)
    println("Parsed args:")
    for (arg, val) in args
        println("$arg  =>  $val")
    end

    qqfly(args["out"], args["title"], args["cutoff"], args["bins"])

end

# running on terminal with one-liner
# zcat pval.gz | awk 'NR>1{print $10}'| qqplot.jl --out test --title "QQ plot on the fly"
if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
# the recommanded is running in julia REPL as follows
# cmd = pipeline(`zcat pval.gz`, `awk 'NR>1{print $10}'`)
# qqfly("test.out", cmd=cmd)

File: qqplot.py
#!/usr/bin/env python3

import sys
import argparse
import numpy as np
import matplotlib.pyplot as plt

from time import localtime, strftime


def _plot(x, y, ax=None, color=None, ablinecolor="r", alpha=0.8, **kwargs):
    """
    Parameters
    ----------
    x, y : array-like, x is the expected, y is the observed

    ax : matplotlib axis, optional

    color : the dots color in the plot, optional

    kwargs : key, value pairings passed to ``plt.scatter()``

    Returns
    -------
    ax : matplotlib Axes object with the plot.
    """
    # Draw the plot and return the Axes
    if ax is None:
        _, ax = plt.subplots(figsize=(6, 6), facecolor="w", edgecolor="k")

    # Get the color from the current color cycle
    if color is None:
        (line,) = ax.plot(0, x.mean())
        color = line.get_color()
        line.remove()

    ax.scatter(x, y, c=color, alpha=alpha, edgecolors="none", **kwargs)
    ax.set_xlim(xmin=x.min(), xmax=1.05 * x.max())
    ax.set_ylim(ymin=y.min())

    if ablinecolor:
        # plot the y=x line
        ax.plot(
            [x.min(), ax.get_xlim()[1]],
            [x.min(), ax.get_xlim()[1]],
            color=ablinecolor,
            linestyle="-",
        )

    return ax


def _group_bins(cutoff, bins):

    if cutoff < 0 or cutoff > 1:
        raise ValueError("`cutoff` must be a float in (0, 1).")
    # bin: [pval, size]
    allbins = [[-1, 0] for _ in range(bins)]
    # keep all the low pvals, which may be significant
    sigp = []
    N = 0
    c = -np.log10(cutoff)
    sys.stdout.write(
        strftime("%a, %d %b %Y %H:%M:%S", localtime())
        + f" : start to process the input!\n"
    )
    try:
        for line in sys.stdin:
            N += 1
            if N % 100000000 == 0:
                sys.stdout.write(
                    strftime("%a, %d %b %Y %H:%M:%S", localtime())
                    + f" : reaching the {N}-th line.\n"
                )
            p = -np.log10(float(line))
            if p > c:
                sigp.append(p)
            else:
                # find the bin where the P should go
                idx = bins - int(np.floor(p / c * bins)) - 1
                idx = 0 if idx == -1 else idx
                allbins[idx][0] = max(allbins[idx][0], p)
                allbins[idx][1] += 1
    except KeyboardInterrupt:
        pass

    sys.stdout.write(
        strftime("%a, %d %b %Y %H:%M:%S", localtime())
        + f" : reached the end! the total number of lines is {N}!\n"
    )
    size = len(sigp)
    obs = sorted(sigp, reverse=True)
    exp = [-np.log10((i + 1 - 0.5) / N) for i in np.arange(size)]

    for bi in allbins:
        if bi[0] != -1:
            obs.append(bi[0])
            exp.append(-np.log10((bi[1] + size - 0.5) / N))
            size += bi[1]
        else:
            pass

    return np.array(obs), np.array(exp)


def qq(
    x=None,
    y=None,
    figname=None,
    cutoff=1e-4,
    bins=1000,
    ax=None,
    title=None,
    color=None,
    alpha=0.8,
    ablinecolor="r",
    dpi=300,
    xlabel=None,
    ylabel=None,
    **kwargs,
):

    if x is None:
        # read data from stdin
        o, e = _group_bins(cutoff, bins)
    else:
        if y is None:
            n = len(x)
            a = 0.5
            o = -np.log10(sorted(x))
            e = -np.log10((np.arange(n, dtype=float) + 1 - a) / (n + 1 - 2 * a))
        else:
            o = x
            e = y
    ax = _plot(e, o, ax=ax, color=color, ablinecolor=ablinecolor, alpha=alpha, **kwargs)
    if xlabel is None:
        xlabel = r"Expected $-log_{10}{(P)}$"
    if ylabel is None:
        ylabel = r"Observed $-log_{10}{(P)}$"
    ax.set_title(title)
    ax.set_xlabel(xlabel)
    ax.set_ylabel(ylabel)
    plt.savefig(figname, bbox_inches="tight", dpi=dpi)
    plt.clf()
    plt.close()

    return ax


def main():

    parser = argparse.ArgumentParser(
        description="QQ plot on the fly! only read data from the pipe!"
    )
    parser.add_argument(
        "-cutoff",
        metavar="FLOAT",
        nargs="?",
        default=1e-4,
        type=float,
        help="pval bigger than this cutoff will be grouped into bins. [%(default)s]",
    )
    parser.add_argument(
        "-bins",
        metavar="INT",
        nargs="?",
        default=1000,
        type=int,
        help="the number of bins to use. [%(default)s]",
    )
    parser.add_argument(
        "-csv",
        default=False,
        action="store_true",
        dest="csv",
        help="read csv (two columns) from stdin. [%(default)s]",
    )
    parser.add_argument(
        "-no",
        default=False,
        action="store_true",
        dest="no",
        help="no grouping. [%(default)s]",
    )
    parser.add_argument("-out", metavar="FILE", help="prefix of output files")
    parser.add_argument("-title", metavar="STRING", help="title of the plot")
    args = parser.parse_args()

    assert args.out is not None, "please specify the file of output."

    f, ax = plt.subplots(figsize=(6, 6), facecolor="w", edgecolor="k")
    if args.csv:

        import pandas as pd

        d = pd.read_csv(sys.stdin, header=0, names=["o", "e"])
        qq(
            x=d["o"].values,
            y=d["e"].values,
            figname=args.out + ".png",
            title=args.title,
            ax=ax,
        )
    elif args.no:
        o = list(map(float, filter(None, (row.rstrip() for row in sys.stdin))))
        qq(
            x = o,
            figname=args.out + ".png",
            cutoff=args.cutoff,
            bins=args.bins,
            title=args.title,
            ax=ax,
        )
    else:
        qq(
            figname=args.out + ".png",
            cutoff=args.cutoff,
            bins=args.bins,
            title=args.title,
            ax=ax,
        )


if __name__ == "__main__":
    main()
